src/v/cloud_storage/partition_recovery_manager.cc

```diff
@@ -406,15 +388,18 @@ partition_downloader::download_log_with_capped_size(
             s.size_bytes,
             dlpart.part_prefix,
             dlpart.dest_prefix);
-          return download_segment_file(s, dlpart).then(
-            [&dloffsets](auto offsets) {
-                if (offsets) {
-                    dloffsets.push_back({
-                      .min_offset = offsets->min_offset,
-                      .max_offset = offsets->max_offset,
-                    });
-                }
-            });
+          return download_segment_file(s, dlpart).then([&](auto offsets) {
+              if (!offsets || offsets->size_bytes == 0u) {
+                  return;
+              }
+              auto [min_offset, max_offset, size] = *offsets;
+              data_found = true;
```

Comment:
it's a bit involved, but it boils down to [offset_translation_layer](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/offset_translation_layer.cc#L44) filtering segments. the full sequence is

partition_downloader::download_segment_file will return size 0 for discarded files
it calls remote::download_segment with a lambda that will use offset_translator::copy_stream to copy or not a segment

Reply:
Ah, I see.  So the change here is that we're downloading segments one by one (instead of in parallel), and using the size after reading through the data to to decide when to stop.

I can see how that is correct, but losing the parallel download of segment is a worry.  Maybe we will want to revisit this in a later PR to reinstate parallel downloads, perhaps downloading at least the requested quantity of raw data, and then proceeding to download further segments in this serial way if we did not get enough user data.

Reply:
yes it would be nice to have parallelism, but before this it wasn't really parallel, because [max_concurrency](https://github.com/redpanda-data/redpanda/blob/dev/src/v/cloud_storage/partition_recovery_manager.h#L90) is set to 1, so there's shouldn't be a noticeable perf regression

Reply:
i can open an issue to track this